File: minWindow.cpp
// Time Complexity : O(n+m) // n lenght of s and m length of t 
// Space Complexity : O(1) // max 26 chars 
// Did this code succesfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// 1. Create frequency map of chars in t; Iterate through s and if char at end pointer found in map, reduce its count
// 2. Once counter is zero implies all chars of t found in window, update min len of substring
// 3. Advance start pointer till one of t's char is no longer in window and continue till end pointer reaches end of s. Return substring or ""
 

class Solution {
public:
    string minWindow(string s, string t) {
        if(s.size()==0 || t.size()==0 || s.size()<t.size())
            return "";
        map<char, int> tracker;
        for(auto c: t)
            tracker[c]++;
        int start=0,end=0,len=INT_MAX,r_start=-1,counter=tracker.size();
        string result;
        while(end<s.size()){
            if(tracker.find(s[end]) != tracker.end()){
                // s[end] found
                tracker[s[end]]--;
                if(tracker[s[end]] == 0) counter--;
            }
            end++;
            while(counter==0){
                if(end-start<len){
                    len=end-start;
                    r_start=start;
                }
                if(tracker.find(s[start]) != tracker.end()){
                    tracker[s[start]]++;
                    if(tracker[s[start]] > 0) counter++;
                }
                start++;
            }
        }
        if(r_start==-1) return "";
        result = s.substr(r_start,len);
        return result;
    }
};

File: min_window.py
# Time complexity - O(s+t) : length of s and t string resp.
# Space Complexity: O(s+t) 
# Did this code successfully run on LeetCode?: Yes
# Problems faced while coding this:None
# Approach : Store all the chars of t string with its frequency. Then traverse on s and maintain a window to record char with frequency in it. 


class Solution:
    def minWindow(self, s, t):
        if not t or not s:
            return ""
        
        required = dict()
        for c in t:
            required[c] = required.get(c, 0) + 1
        
        left, right = 0, 0
        formed = 0
        window_counts = {}
        p1, p2 = -1, -1
        min_len = float('inf')
        while right < len(s):
            window_counts[s[right]] = window_counts.get(s[right], 0) + 1
            if s[right] in required and window_counts[s[right]] == required[s[right]]:
                formed += 1
            while left <= right and formed == len(required):
                if right - left + 1 < min_len:
                    min_len = right - left + 1
                    p1 = left
                    p2 = right
                window_counts[s[left]] -= 1
                if s[left] in required and window_counts[s[left]] < required[s[left]]:
                    formed -= 1
                left += 1    
            
            right += 1
            
        return "" if min_len == float("inf") else s[p1:p2 + 1]